src/p2panda-api/instance.ts

```diff
@@ -0,0 +1,99 @@
+import { Session } from '~/p2panda-api';
+import { Fields } from '~/p2panda-api/types';
+
+import type { Resolved } from '~/typescript/helpers';
+
+type InstanceArgs = {
+  // @ts-expect requires types exported from rust
+  keyPair: Resolved<Session['p2panda']['KeyPair']>;
+  schema: string;
+  session: Session;
+};
+
+/**
+ * Returns a message fields instance for the given field contents and schema
+ */
+const getMessageFields = async (
+  session: Session,
+  _schema: string,
+  fields: Fields,
+) => {
+  const { MessageFields } = await session.loadWasm();
+
+  const messageFields = new MessageFields();
+  for (const fieldName of Object.keys(fields)) {
+    const fieldType = Object.keys(fields[fieldName])[0];
+    messageFields.add(fieldName, fields[fieldName][fieldType]);
+  }
+  return messageFields;
+};
```

Comment:
I don't think this preserves the order in which fields are added.

Reply:
The ordering is automatically done by the underlying BTreeMap, so that should be fine